File: app/widgets.py
import re

import flet as ft


def initials(name):
    return name.strip()[:1].upper() if name.strip() else '?'


def fmt_time(ts):
    import datetime
    return datetime.datetime.fromtimestamp(ts / 1000).strftime('%H:%M')


def fmt_day(ts, tr):
    import datetime
    d = datetime.datetime.fromtimestamp(ts / 1000)
    today = datetime.date.today()
    if d.date() == today:
        return tr('today')
    if d.date() == today - datetime.timedelta(days=1):
        return tr('yesterday')
    return d.strftime('%d %B').lstrip('0')


def avatar(chat, size=50, font_size=None, p=None):
    content = chat.get('icon') or initials(chat.get('name', ''))
    return ft.Container(
        width=size,
        height=size,
        border_radius=size / 2,
        bgcolor=chat.get('color', '#2AABEE'),
        alignment=ft.Alignment.CENTER,
        content=ft.Text(
            content,
            color=ft.Colors.WHITE,
            size=font_size or (size * 0.4 if chat.get('icon') else size * 0.38),
            weight=ft.FontWeight.W_600,
        ),
    )


def day_pill(text, p):
    return ft.Container(
        bgcolor=ft.Colors.with_opacity(0.16, '#7F8C98'),
        border_radius=10,
        padding=ft.Padding.symmetric(horizontal=11, vertical=4),
        margin=ft.margin.symmetric(vertical=8),
        content=ft.Text(text, size=11.5, color=p.text_soft, weight=ft.FontWeight.W_500),
    )


def hashtag_text(text, p, size=14.5):
    spans = []
    last = 0
    for m in re.finditer(r'#[\wа-яА-ЯёЁ]+', text):
        if m.start() > last:
            spans.append(ft.TextSpan(text[last:m.start()]))
        spans.append(ft.TextSpan(
            m.group(0),
            style=ft.TextStyle(color=p.accent_dk, weight=ft.FontWeight.W_600),
        ))
        last = m.end()
    if last < len(text):
        spans.append(ft.TextSpan(text[last:]))
    if not spans:
        spans.append(ft.TextSpan(text))
    return ft.Text(spans=spans, size=size, color=p.text, selectable=True)


def chat_preview(entry, tr):
    t = entry.get('type')
    if t == 'text':
        return entry.get('text', '')[:60]
    if t == 'image':
        return tr('photo')
    if t == 'audio':
        return tr('voice_preview', entry.get('duration', 0))
    if t == 'video':
        return tr('video_preview')
    if t == 'todo':
        items = entry.get('items') or []
        done = sum(1 for i in items if i.get('done'))
        return tr('todo_progress', done, len(items))
    return ''


def build_chat_row(chat, state, p, on_open, tr):
    entries = state.entries_for(chat['id'])
    last = entries[-1] if entries else None
    return ft.GestureDetector(
        on_tap=lambda e: on_open(chat['id']),
        content=ft.Container(
            bgcolor=p.bg_list,
            padding=ft.Padding.symmetric(horizontal=16, vertical=10),
            border=ft.Border.only(bottom=ft.BorderSide(1, p.divider)),
            content=ft.Row(
                spacing=12,
                controls=[
                    avatar(chat),
                    ft.Column(
                        expand=True,
                        spacing=2,
                        controls=[
                            ft.Row(
                                controls=[
                                    ft.Text(chat['name'], size=15.5, weight=ft.FontWeight.W_600, color=p.text, expand=True, overflow=ft.TextOverflow.ELLIPSIS),
                                    ft.Text(fmt_time(last['ts']) if last else '', size=12, color=p.text_faint),
                                ],
                            ),
                            ft.Row(
                                controls=[
                                    ft.Text(chat_preview(last, tr) if last else tr('no_entries'), size=13.5, color=p.text_soft, expand=True, overflow=ft.TextOverflow.ELLIPSIS, max_lines=1),
                                    ft.Container(
                                        bgcolor=p.accent,
                                        border_radius=10,
                                        padding=ft.Padding.symmetric(horizontal=5, vertical=2),
                                        content=ft.Text(str(len(entries)), size=11, color=ft.Colors.WHITE, weight=ft.FontWeight.W_600),
                                    ) if entries else ft.Container(),
                                ],
                            ),
                        ],
                    ),
                ],
            ),
        ),
    )


def build_forward_row(chat, p, on_pick):
    return ft.GestureDetector(
        on_tap=lambda e: on_pick(chat),
        content=ft.Container(
            padding=ft.Padding.symmetric(vertical=8, horizontal=4),
            content=ft.Row(
                spacing=10,
                controls=[
                    avatar(chat, size=38),
                    ft.Text(chat['name'], size=15, weight=ft.FontWeight.W_600, color=p.text, expand=True),
                ],
            ),
        ),
    )

File: lib/src/widgets.dart
import 'package:flutter/material.dart';

import 'app_model.dart';
import 'models.dart';
import 'theme.dart';

Color _hex(String hex) => colorFromHex(hex);

class ChatAvatar extends StatelessWidget {
  const ChatAvatar({super.key, required this.chat, this.size = 46, this.iconSize = 22});

  final Chat chat;
  final double size;
  final double iconSize;

  @override
  Widget build(BuildContext context) {
    final letter = chat.name.isNotEmpty ? chat.name[0].toUpperCase() : '?';
    final bg = _hex(chat.color);
    return Container(
      width: size,
      height: size,
      decoration: BoxDecoration(
        color: bg,
        shape: BoxShape.circle,
        boxShadow: [
          BoxShadow(
            color: bg.withValues(alpha: 0.22),
            blurRadius: 8,
            offset: const Offset(0, 2),
          ),
        ],
      ),
      alignment: Alignment.center,
      child: chat.icon != null && chat.icon!.isNotEmpty
          ? Text(chat.icon!, style: TextStyle(fontSize: iconSize))
          : Text(letter, style: TextStyle(fontSize: iconSize * 0.72, color: Colors.white, fontWeight: FontWeight.w800, letterSpacing: 0.2)),
    );
  }
}

class DayPill extends StatelessWidget {
  const DayPill({super.key, required this.label, required this.p});

  final String label;
  final Palette p;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 5),
        decoration: BoxDecoration(
          color: p.bgChat,
          borderRadius: BorderRadius.circular(TNRadii.pill),
          border: Border.all(color: p.divider.withValues(alpha: 0.5)),
        ),
        child: Text(label, style: TextStyle(fontSize: 11.5, fontWeight: FontWeight.w600, color: p.textFaint, letterSpacing: 0.2)),
      ),
    );
  }
}

class ChatRow extends StatelessWidget {
  const ChatRow({
    super.key,
    required this.chat,
    required this.p,
    required this.preview,
    required this.time,
    required this.onTap,
    this.onLongPress,
    this.highlight = false,
  });

  final Chat chat;
  final Palette p;
  final String preview;
  final String time;
  final VoidCallback onTap;
  final VoidCallback? onLongPress;
  final bool highlight;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: highlight ? p.rowActive : Colors.transparent,
      child: InkWell(
        onTap: onTap,
        onLongPress: onLongPress,
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 11),
          child: Row(
            children: [
              ChatAvatar(chat: chat),
              const SizedBox(width: 12),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Row(
                      children: [
                        if (chat.pinned) ...[
                          Container(
                            padding: const EdgeInsets.all(3),
                            decoration: BoxDecoration(color: p.accent.withValues(alpha: 0.12), shape: BoxShape.circle),
                            child: Icon(Icons.push_pin, size: 10, color: p.accent),
                          ),
                          const SizedBox(width: 6),
                        ],
                        Text(chatKinds.firstWhere((k) => k.$1 == chat.kind, orElse: () => ('note', '📝')).$2, style: const TextStyle(fontSize: 13)),
                        const SizedBox(width: 5),
                        Expanded(
                          child: Text(chat.name,
                              maxLines: 1,
                              overflow: TextOverflow.ellipsis,
                              style: TNTypography.chatTitle.copyWith(color: p.text)),
                        ),
                        if (time.isNotEmpty)
                          Container(
                            padding: const EdgeInsets.symmetric(horizontal: 6, vertical: 2),
                            decoration: BoxDecoration(
                              color: p.bgChat,
                              borderRadius: BorderRadius.circular(TNRadii.pill),
                            ),
                            child: Text(time, style: TNTypography.time.copyWith(color: p.textFaint)),
                          ),
                      ],
                    ),
                    const SizedBox(height: 4),
                    Text(preview,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: TNTypography.chatPreview.copyWith(color: p.textSoft)),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class SearchResultRow extends StatelessWidget {
  const SearchResultRow({
    super.key,
    required this.chat,
    required this.p,
    required this.snippet,
    required this.onTap,
  });

  final Chat chat;
  final Palette p;
  final String snippet;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    return Material(
      color: Colors.transparent,
      child: InkWell(
        onTap: onTap,
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 11),
          child: Row(
            children: [
              ChatAvatar(chat: chat, size: 38, iconSize: 18),
              const SizedBox(width: 10),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(chat.name,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: TextStyle(fontSize: 14, fontWeight: FontWeight.w600, color: p.text)),
                    const SizedBox(height: 3),
                    Text(snippet,
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: TextStyle(fontSize: 12.5, color: p.textSoft, height: 1.3)),
                  ],
                ),
              ),
              Icon(Icons.chevron_right, size: 16, color: p.textFaint.withValues(alpha: 0.6)),
            ],
          ),
        ),
      ),
    );
  }
}

String snippetFor(Entry e, String q, String Function(String, [List<String>?]) tr) {
  final preview = entryPreview(e, tr);
  final low = preview.toLowerCase();
  final qi = low.indexOf(q.toLowerCase());
  if (qi >= 0) {
    final start = qi > 20 ? qi - 20 : 0;
    return '…${preview.substring(start, qi + q.length + 40 > preview.length ? preview.length : qi + q.length + 40)}';
  }
  return preview;
}
